File: domain/role.go
package domain

import (
	"context"
	"time"
)

// Role represents a custom role in the system
type Role struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Sequence  int       `json:"sequence"`
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	MenusIds  []string  `json:"menus"`
}

// CreateRoleRequest represents a request to create a new role
type CreateRoleRequest struct {
	Name     string   `json:"name" binding:"required,max=100"`
	Sequence int      `json:"sequence"`
	Status   string   `json:"status,omitempty"`
	MenuIDs  []string `json:"menu_ids,omitempty"` // Menu permissions managed through Casbin
}

// UpdateRoleRequest represents a request to update an existing role
type UpdateRoleRequest struct {
	Name     *string  `json:"name,omitempty"`
	Sequence *int     `json:"sequence,omitempty"`
	Status   *string  `json:"status,omitempty"`
	MenuIDs  []string `json:"menu_ids,omitempty"` // Menu permissions managed through Casbin
}

// RolePagedResult represents paginated role query results
type RolePagedResult = PagedResult[*Role]

// RoleFilters represents filters for role queries
type RoleFilters struct {
	Status string `json:"status,omitempty"`
	Name   string `json:"name,omitempty"`
}

// RoleUseCase defines the interface for role business logic
type RoleUseCase interface {
	Create(c context.Context, request *CreateRoleRequest) error
	Fetch(c context.Context) ([]*Role, error)
	FetchPaged(c context.Context, params QueryParams) (*RolePagedResult, error)
	GetByID(c context.Context, id string) (*Role, error)
	Update(c context.Context, id string, request *UpdateRoleRequest) (*Role, error)
	Delete(c context.Context, id string) error
}

// RoleRepository defines the interface for role data operations
type RoleRepository interface {
	Create(c context.Context, role *Role) error
	Fetch(c context.Context) ([]*Role, error)
	FetchPaged(c context.Context, params QueryParams) (*RolePagedResult, error)
	GetByID(c context.Context, id string) (*Role, error)
	GetByName(c context.Context, name string) (*Role, error)
	Update(c context.Context, role *Role) error
	Delete(c context.Context, id string) error
	Assign(c context.Context, uid, rid string) error
	Revoke(c context.Context, uid, rid string) error
}

File: web/src/types/role.ts
// 角色信息 (对应后端RoleInfo)
export interface RoleInfo {
  id: string // 角色ID
  name: string // 角色显示名称
  type: string // 角色类型
}

// Role Management
export interface Role {
  id: string
  name: string
  sequence: number
  status: 'active' | 'inactive'
  created_at: Date
  updated_at: Date
  menu_ids?: string[]
}

export interface CreateRoleRequest {
  name: string
  sequence: number
  status?: 'active' | 'inactive'
  menu_ids?: string[]
}

export interface UpdateRoleRequest {
  name?: string
  sequence?: number
  status?: 'active' | 'inactive'
  menu_ids?: string[]
}

export interface RolePagedResult {
  list: Role[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface AssignRoleRequest {
  user_id: string
  role_id: string
}
